librosa/beat.py

```diff
@@ -320,30 +354,24 @@ def tempo(y=None, sr=22050, onset_envelope=None, hop_length=512, start_bpm=120,
     bpms = core.tempo_frequencies(tg.shape[0], hop_length=hop_length, sr=sr)
 
     # Weight the autocorrelation by a log-normal distribution
-    prior = np.exp(-0.5 * ((np.log2(bpms) - np.log2(start_bpm)) / std_bpm)**2)
+    if prior is None:
+        logprior = -0.5 * ((np.log2(bpms) - np.log2(start_bpm)) / std_bpm)**2
+    else:
+        logprior = prior.logpdf(bpms)
 
     # Kill everything above the max tempo
     if max_tempo is not None:
         max_idx = np.argmax(bpms < max_tempo)
-        prior[:max_idx] = 0
-
-    # Really, instead of multiplying by the prior, we should set up a
-    # probabilistic model for tempo and add log-probabilities.
-    # This would give us a chance to recover from null signals and
-    # rely on the prior.
-    # it would also make time aggregation much more natural
+        logprior[:max_idx] = -np.inf
 
     # Get the maximum, weighted by the prior
-    best_period = np.argmax(tg * prior[:, np.newaxis], axis=0)
+    best_period = np.argmax(np.log(tg + 1e-10) + logprior[:, np.newaxis], axis=0)
```

Comment:
i'm not a big fan of `np.log(x + 1e-10)`
consider replacing by `np.log1p(1e10 * x)`, which is the same up to an additive constant (not affecting the argmax), and more numerically stable for small `x`.